experimental/traffic-portal/src/app/core/certs/cert-viewer/cert-viewer.component.ts

```diff
@@ -0,0 +1,187 @@
+	}
+
+	/**
+	 * newCert creates a cert from an input string.
+	 *
+	 * @param input The text to read as a cert
+	 * @private
+	 * @returns Resultant Cert
+	 */
+	private newCert(input: string): AugmentedCertificate {
+		try {
+			return pki.certificateFromPem(input) as AugmentedCertificate;
+		} catch (e) {
+			console.error(`ran into issue creating certificate from input ${input}`, e);
+			return NULL_CERT;
+		}
+	}
+
+	/**
+	 * process takes the Cert Chain text input and parses it.
+	 *
+	 * @param uploaded if the certificate was uploaded by the client.
+	 */
+	public process(uploaded: boolean = false): void {
+		this.inputCert = this.inputCert.replace(/\r\n/g, "\n");
+		const parts = this.inputCert.split("-\n-");
+		const certs = new Array<AugmentedCertificate>(parts.length);
+		for(let i = 1; i < parts.length; ++i) {
+			parts[i-1] += "-";
+			parts[i] = `-${parts[i]}`;
+			certs[i-1] = this.newCert(parts[i - 1]);
+		}
+		certs[certs.length-1] = this.newCert(parts[parts.length - 1]);
+		const assignType = (c: AugmentedCertificate, i: number): void => {
+			if(c.parseError) {
+				return;
+			}
+			if (i === 0) {
+				c.type = "Root";
+			} else if (i === certs.length - 1) {
+				c.type = "Client";
+			} else {
+				c.type = "Intermediate";
+			}
+		};
+		const chain = this.reOrderRootFirst(certs);
+		chain.forEach(assignType);
+		this.certChain = chain;
+
+		if(this.matTab && uploaded) {
+			this.matTab.selectedIndex = 1;
+		}
+	}
+
+	/**
+	 * reOrderRootFirst sorts a cert chain with the root being first if possible.
+	 *
+	 * @param certs The list of certs to reorder
+	 * @returns The processed certs
+	 */
+	public reOrderRootFirst(certs: Array<AugmentedCertificate>): Array<AugmentedCertificate> {
+		let rootFirst = false;
+		let invalid = false;
+		for(let i = 1; i < certs.length; ++i){
+			const first = certs[i-1];
+			const next = certs[i];
+			if(first.parseError) {
+				invalid = true;
+				continue;
+			} else if (next.parseError) {
+				invalid = true;
+				continue;
+			}
+			if (first.issued(next)) {
+				rootFirst = true;
+			} else if (next.issued(first)) {
+				rootFirst = false;
+			} else {
+				invalid = true;
+				console.error(`Cert chain is invalid, cert ${i-1} and ${i} are not related`);
+			}
+		}
+
+		if (certs.length === 1) {
+			if (certs[0].parseError) {
+				invalid = true;
+			} else {
+				this.certOrder = "Single";
+				return certs;
+			}
+		}
+		if (invalid) {
+			this.certOrder = "Unknown";
+			return certs;
+		}
+
+		if(rootFirst) {
+			this.certOrder = "Root -> Client";
+			return certs;
+		}
+		this.certOrder = "Client -> Root";
+		certs = certs.reverse();
+		return certs;
+	}
+
+	/**
+	 * Checks if we are a DS cert or any user provided cert.
+	 */
+	public async ngOnInit(): Promise<void> {
+		const ID = this.route.snapshot.paramMap.get("xmlId");
+		if (ID === null) {
+			this.dsCert = false;
+			return;
+		}
+		this.cert = await this.dsAPI.getSSLKeys(ID);
```

Comment:
if I navigate to the SSL certificate inspection page for a non-existent DS (e.g. if I click on an old link for a DS that has since been deleted), then `dsCert` will be `true`, but the actual page shown is the one for manual input and inspection - _not_ inspecting a DS certificate. Also, the URL will remain showing me the nonexistent XMLID - it should instead replace the routing state to the non-DS certificate viewer URL.